Restructure Flux Linkage and Inductance Analyzers

dmnewman3 opened the pull request:
# Issue
Closes #296 

# Summary
This PR does the following:
- separates existing flux linkage and inductance analyzers
- enables the use of more than 3-phase machines
- provides the full alpha-beta and d-q reference frame inductances including cross-coupled inductances
- separates the documentation and .rst files for each.

closes #296

dmnewman3 committed:
Separate inductance analyzer and flux linkage analyzer

dmnewman3 committed:
rename to proper convention

dmnewman3 committed:
Merge branch 'develop' into user/dmnewman3/restructure_inductance_analyzer

dmnewman3 committed:
separate flux linkage and inductance analyzers

dmnewman3 committed:
fix grammar and spelling of .rst files

dmnewman3 committed:
fix links

dmnewman3 committed:
restructure flux_linkage analyzer and documentation

dmnewman3 committed:
restructure inductance analyzer before debugging

dmnewman3 committed:
fix array

dmnewman3 committed:
Finalize documentation of inductance analyzer

dmnewman3 committed:
Merge branch 'develop' into user/dmnewman3/restructure_inductance_analyzer

# Conflicts:
#	docs/source/EM_analyzers/inductance_analyzer.rst
#	examples/mach_eval_examples/SynR_eval/inductance_step.py
#	mach_eval/analyzers/electromagnetic/flux_linkage_analyzer_config.py

dmnewman3 committed:
update language of flux analyzer .rst file

wchan29 reviewed (dismissed):
@dmnewman3 Here is my Level 1 review summary:

Does the code run without error and produce the expected result? Yes
Does the code comply with the [code guidelines](https://emach.readthedocs.io/en/latest/code.html)? Yes
Does the code documentation comply with the [documentation guidelines](https://emach.readthedocs.io/en/latest/documentation.html)? Yes
Is the writing, grammar, and syntax correct and clear? Yes
Is the changeset compliant with the eMach architecture? Yes
Does this review consider whether this physics are accurate? Not sure, but seems good.
Is PR approved to Level 2? Yes

dmnewman3 committed:
remove references to a 3-phase machine

elsevers commented:
@dmnewman3 -- thanks! I have asked @noguchi-takahiro to do a review on this. Once you get an approval from him, I'll go ahead and do a review as well.

dmnewman3 committed:
add code to specify geometry to avoid naming issues
- main mach_cad file
- tutorial 5 file

noguchi-takahiro commented on examples/mach_eval_examples/SynR_eval/SynR_flux_linkage_step.py:
```diff
@@ -0,0 +1,72 @@
+        print("path = ", state_out.conditions.path)
+        print("study_name = ", state_out.conditions.study_name)
+        print("I_hat = ", state_out.conditions.I_hat, " A")
+        print("rotor_angle = ", state_out.conditions.rotor_angle, " deg")
```
This is an optional change but when we print the rotor_angle array, it would be better if we have theese vaulues without mention of "array" itself. Could you consider if this change is acceptable for you?

```suggestion
        print("rotor_angle = ", state_out.conditions.rotor_angle[0], " deg")
```

I am expecting this result:
![image](https://github.com/Severson-Group/eMach/assets/114006024/e36e9374-6c85-451d-87c8-c5783a3b11b3)

dmnewman3 replied:
Thank you @noguchi-takahiro, I have made this change.

noguchi-takahiro reviewed (commented):
Thank you @dmnewman3. Your flux and inductance analyzer works well. I have super small request regarding the printing the rotor_angle. Please see the comment below and consider to update it.

dmnewman3 committed:
update output print statement per Takahiro's feedback

noguchi-takahiro reviewed (dismissed):
Thank you @dmnewman3. It looks good to me.

Here is my Level 1 review summary:

Does the code run without error and produce the expected result? Yes
Does the code comply with the [code guidelines](https://emach.readthedocs.io/en/latest/code.html)? Yes
Does the code documentation comply with the [documentation guidelines](https://emach.readthedocs.io/en/latest/documentation.html)? Yes
Is the writing, grammar, and syntax correct and clear? Yes
Is the changeset compliant with the eMach architecture? Yes
Does this review consider whether this physics are accurate? Yes
Is PR approved to Level 2? Yes

@elsevers I approve this PR. Could you review this?

noguchi-takahiro committed:
Fix Stator ID in imag_2d

noguchi-takahiro committed:
Uncomment codes to rename parts

dmnewman3 committed:
make changes based on feedback:
- remove references to version 21.1
- remove extra stator id naming line

dmnewman3 committed:
fix SynR and flux linkage analyzer for JMAG23

dmnewman3 committed:
Merge branch 'feature/fix_geometry_naming' into user/dmnewman3/restructure_inductance_analyzer

elsevers reviewed (changes requested):
Partial review here. I am comparing the plan in [issue #296](https://github.com/Severson-Group/eMach/issues/296#issuecomment-2004995071) to the documentation I am reading for the flux linkage analyzer and think we aren't quite on the right track.
1. Flux linkage analyzer input 1: instead of taking in a SynR machine object, instead take in a JMAG tool handle to a JMAG project that is open and already loaded with the machine and fully configured. It could be a SynR machine, or a PM machine, or something else entirely -- doesn't matter.
2. Flux linkage analyzer input 2: total number of phases
3. Flux linkage analyzer input 3: name of each phase
4. Flux linkage analyzer input 4: list consisting of objects of our standard operating point class to calculate flux linkages at.

The analyzer then calculates the flux linkages for all phases at each operating point. 

One thing I am not sure of: does our operating point class give us capability to specify each phase current individually? 

Takahiro and I will discuss a bit more later today and revise this description.

dmnewman3 committed:
update .rst file documentation based on Eric's feedback - part 1

dmnewman3 committed:
restructure analyzer files according to 5/31 feedback

dmnewman3 committed:
update documentation according to 5/31 meeting

dmnewman3 commented:
Hi @noguchi-takahiro, I've updated the analyzers and incorporated @elsevers feedback. I think I've got it to a point where I can have you look over it. It have it to a point where everything is structured as Eric requested and it is fully functional with the example file I pushed to the repo. I've also updated the documentation quite a bit.

elsevers reviewed (changes requested):
Change requests for flux linkage analyzer and problem
- problem: rename to `Flux_Linkage_Problem` to `FluxLinkageJMAG_Problem`
- problem: Use the `toolJmag = JMAG.JmagDesigner()` instead of `app` and `model`
- problem + analyzer: let's eliminate using CSV files to exchange information. Instead, we'll save flux linkage data to an object. In the problem class, see if you can prevent JMAG from storing the data onto the hard drive. Just retrieve it directly from JMAG via a function call.
- problem: `ratedcurrent` should be renamed to `rated_current`
- analyzer: should not be referring to JMAG specific information from the problem class. I.e., remove the analyzer using app or model.
- problem: remove the requirement that a user needs to make the mesh prior to creating the problem. Our theory is that creating the mesh as a distinct step is unnecessary and that just calling the `run` function will take care of this. The model should be pre-configured with the appropriate mesh settings prior to creating the problem object.

dmnewman3 committed:
update files according to 6/10 feedback from Eric for PR 343

dmnewman3 commented:
Thanks for the feedback @elsevers and @noguchi-takahiro, I have made the changes requested:

- [x] renamed problem and analyzer to `FluxLinkageJMAG_Problem` and `FluxLinkageJMAG_Analyzer`
- [x] restructured to have `toolJmag` as the input instead of `app` and `model`
- [x] eliminated the need for csv files and restructured the`InductanceProblem` and `InductanceAnalyzer` accordingly
- [x] renamed `ratedcurrent` to `rated_current`
- [x] moved all specific JMAG references to problem class in flux linkage analyzer
- [x] removed the need to include the mesh in the initial FEA file

Please re-review and provide feedback as necessary.

elsevers commented on mach_eval/analyzers/electromagnetic/flux_linkage_analyzer.py:
```diff
@@ -0,0 +1,173 @@
+                            self.results_filepath + self.study_name + "_flux_of_fem_coil_phase_%s.csv" % problem.phase_names[i-1])
+                else:
+                    os.rename(self.results_filepath + self.study_name + "_flux_of_fem_coil.csv", 
+                            self.results_filepath + self.study_name + "_flux_of_fem_coil_phase_%s.csv" % problem.phase_names[i-1])
```
@dmnewman3, I just took a very quick look and I see that this is still using CSV files.

I thought we decided not to use CSV files / or write anything to the hard drive. I.e.:
> problem + analyzer: let's eliminate using CSV files to exchange information. Instead, we'll save flux linkage data to an object. In the problem class, see if you can prevent JMAG from storing the data onto the hard drive. Just retrieve it directly from JMAG via a function call.

Can you implement this?

dmnewman3 replied:
@elsevers , I looked at all of the other analyzers that utilitze JMAG outputs and they all involve writing to and then using .csv files, back to and including the original BSPM analyzer that Barath made. I tried to go into JMAG and see if there was a way around this, but the only manner in which I could generate data was outputting it to a file, csv or otherwise. 

What I changed regarding the use of the .csv files in this newest update was to write to the .csv and then use it's contents to create an object all within the flux linkage analyzer. After this update, we are no longer passing the .csv or it's filepath location between analyzers. This seemed the most logical solution based on our discussion and what I found using JMAG. Please let me know your thoughts here.

elsevers replied:
Okay, if we need to use a CSV file, please keep it contained to happening within the problem class. The analyzer shouldn't have any knowledge of the CSV file existing. 

Remember, this analyzer shouldn't know we're dealing with JMAG. Someone should be able to write a future problem class that this analyzer could use that might use FEMM or Maxwell or ...

dmnewman3 replied:
Ok @elsevers I just updated to move all references of the csv files to the problem class. Please let me know if any further adjustments are necessary.

elsevers reviewed (changes requested):
@dmnewman3, quick review on this caught a problem -- can you take a look?

dmnewman3 committed:
update analyzer to eliminate references to the csv file(s) in the analyzer and move them to the problem class

elsevers commented on mach_eval/analyzers/electromagnetic/flux_linkage_analyzer.py:
```diff
@@ -0,0 +1,171 @@
+            ################################################################
+            problem.run_study(self.study_name, study, i, clock_time())
+        
+        ####################################################
```
@dmnewman3, this logic doesn't seem to match what we outlined [in #296](https://github.com/Severson-Group/eMach/issues/296#issuecomment-2142904755)

Your code: 
1. sets the current for a case
2. runs FEA
3. goes back to 1 for the next case, and does this for as many times as there are currents
4. extracts the results.

This is a "serial" or sequential approach, which does not allow a clever problem class to parallelize the operation.

[In #296](https://github.com/Severson-Group/eMach/issues/296#issuecomment-2142904755) we specify an approach that allows a clever, future, problem class to parallelize the operation:
1. set the currents up for all cases
2. tell the problem to solve all cases
3. extract the results

Can you please update your approach to match what we specify in [#296](https://github.com/Severson-Group/eMach/issues/296#issuecomment-2142904755)?

dmnewman3 replied:
@elsevers I see, I have updated accordingly to what was described in #296. I have also tried to comment out sections of the code so it's easier to understand. Please let me know what you think after reviewing this.

elsevers reviewed (changes requested):
Thanks @dmnewman3. See my comment on revisions needed to implement the approach we outline in [#296](https://github.com/Severson-Group/eMach/issues/296#issuecomment-2142904755)

dmnewman3 committed:
Make changes to the following for using parallel cases:
- requirements of model in .rst file
- flux_linkage analyzer
- inductance analyzer
- example JMAG file

elsevers commented on examples/mach_eval_examples/fluxlinkage_inductance_eval/fluxlinkage_inductance_evaluator.py:
```diff
@@ -0,0 +1,103 @@
+
+from mach_cad.tools import jmag as JMAG
+
+filepath = "C:/Users/dante/Documents/github_repos/eMach/examples/mach_eval_examples/fluxlinkage_inductance_eval"
```
This hard coded path won't work for other people. Can you please use relative paths?

elsevers commented on examples/mach_eval_examples/fluxlinkage_inductance_eval/fluxlinkage_inductance_evaluator.py:
```diff
@@ -0,0 +1,103 @@
+import os
```
When I try to run this example file, I encounter this error:

![image](https://github.com/Severson-Group/eMach/assets/7421348/9ccae146-43d6-4a58-b870-f1ea9e71062b)

dmnewman3 replied:
Thanks @elsevers,

I have never seen this issue, but I have implemented a save command right after opening so hopefully that fixes your problem. I have created the file using JMAG21 so there should be no problem opening it in a more recent version. Please do let me know if this most-recent push fixes this.

dmnewman3 replied:
According to @noguchi-takahiro and @dmnewman3, this appears to be a bug with JMAGv22. This error did not pop up when using JMAGv21 or JMAGv23.

elsevers commented on mach_eval/analyzers/electromagnetic/flux_linkage_analyzer.py:
```diff
@@ -0,0 +1,141 @@
+        self.phase_names = phase_names
+        self.rated_current = rated_current
+    
+    def add_em_study(self, study_name):
```
Having the analyzer call this function doesn't seem right. The analyzer shouldn't need to be providing a study name, or even thinking about making studies.

But this could be an internal helper function that the problem class's `run()` function uses.

elsevers commented on mach_eval/analyzers/electromagnetic/flux_linkage_analyzer.py:
```diff
@@ -0,0 +1,141 @@
+        return study
+
+    
+    def del_op_pts(self, study):
```
Let's rename to `delete_operating_points(self)` 
--> and no need to bring `study` in

elsevers commented on mach_eval/analyzers/electromagnetic/flux_linkage_analyzer.py:
```diff
@@ -0,0 +1,141 @@
+            self.app.GetModel(0).GetStudy(0).GetDesignTable().AddCase()
+            self.app.GetModel(0).GetStudy(0).GetDesignTable().SetValue(0, i, 0)
+
+    def new_op_pt(self, I_hat, i):
```
Let's rename to `new_operating_point(self)` and have it return an index for the new operating point (like my pseudocode does [here](https://github.com/Severson-Group/eMach/issues/296#issuecomment-2142904755))

elsevers commented on mach_eval/analyzers/electromagnetic/flux_linkage_analyzer.py:
```diff
@@ -0,0 +1,141 @@
+
+    def new_op_pt(self, I_hat, i):
+
+        # Setting current values of single current source depending on phase excitation
```
Create a new function called
`set_phase_current(op, phase_num, val)` that set's the current current of `phase_num` to `val` of operating point `op`, like my psuedo code does [here](https://github.com/Severson-Group/eMach/issues/296#issuecomment-2142904755)

elsevers commented on mach_eval/analyzers/electromagnetic/flux_linkage_analyzer.py:
```diff
@@ -0,0 +1,141 @@
+            self.app.GetModel(0).GetStudy(0).GetDesignTable().SetValue(i, i-1, I_hat)
+
+
+    def run_all_studies(self, study, toc):
```
rename this to `run_analyses(self)` --> it should not need to take in `study` or `toc`

elsevers commented on mach_eval/analyzers/electromagnetic/flux_linkage_analyzer.py:
```diff
@@ -0,0 +1,141 @@
+        
+        # Write output data of analyzer
+        fea_data = {
+                "current_peak": self.rated_current,
```
This should be a vector of the phase currents

elsevers reviewed (changes requested):
Thanks @dmnewman3. I left you some items to work on as in-line annotations.

I think we should be including an abstract base class for that the flux linkage analyzer expects of the flux linkage problem class (abstract functions of `delete_operating_points`, `new_operating_point`, `set_phase_current`, and  `run_analyses`) and then have your problem class implement that abstract base class. This approach is used extensively in other parts of the codebase. If you are familiar with how to do this, go ahead and try. Otherwise, please work on addressing the more concrete items I have identified.

dmnewman3 committed:
Make changes on Eric's 6/17 feedback including:
- remove hard-coded references
- save file before executing
- redefine add_em_study and remove it as analyzer call
- rename and restructure del_op_pts, new_op_pt, run_all_cases
- add set_phase_currents
- set output as vector of phase currents
- update documentation

dmnewman3 commented:
@elsevers I have implemented your feedback, but I don't understand how what we're doing now is not implementing the abstract base classes you are referring to. The pattern we are doing seem to be exactly how is laid out in the guidelines and how other analyzers are structured. I would like to prioritize getting this over the finish line and then if time allows we can improve upon it from there. Either that or maybe I have the wrong idea of what you're referring to.

Either way, I am ready for a new review.
